### ISSUE TITLE
How to estimate the variance?

### ISSUE THREAD

**Issue by xuehy** [alexgkendall/multitaskvision#5]
Very interesting paper! I have a question after reading your paper several times. How do you estimate the uncertainty(variance) of each task?

**Comment by JulienSiems** [comment]
The paper uses homoscedastic uncertainty to weigh the losses, which means that the uncertainty is independent of the input data. For clarification take a look at this [blog post](https://alexgkendall.com/computer_vision/bayesian_deep_learning_for_safe_ai/) and maybe this presentation from a [seminar page 25](https://lmb.informatik.uni-freiburg.de/lectures/seminar_brox/seminar_ss17/uncertainty_slides.pdf). 

You would initialize a variable s = log σ² (for a possible initialization value for s check out [the supplementary material](https://arxiv.org/pdf/1705.07115.pdf)). If you have a squared error loss in your loss function (L(W)) you could wrap it like this (see equation 5 in the paper):

L_uncertainty(W) = 1/2 * exp(-s) * L(W) + 1/2 * s

Did I understand that correctly @alexgkendall?

**Comment by cptay** [comment]
Hi JulienSiems,

Wonder if I can ask you a question, since we hear nothing from the author? I am calculating the homoscedastic uncertainty for classification task, and my formula is

L_uncertainty(W) = exp(-s) * L(W) + 1/2 * s

with L(W) as the cross entropy loss of my task. Is that your understanding?

BTW, thanks for the link to the paper which tries to prevent negative loss terms. Wonder if it leads to better performance?

**Comment by JulienSiems** [comment]
Hi @cptay, 
Yes that is how I understood it and it has worked so far. It also sounded to me like the paper I linked understood it like that. I haven't had a chance to try the modified log term from the paper though and can't say yet whether it gives you better performance.

Would be great if you could shed a light on this @alexgkendall?

**Comment by cptay** [comment]
Hi @JulienSiems,

Thanks for the fast response. I think I will give the modified regularization term a try.

But isn't the regularization term used by Alex = 'log σ', instead of  'log σ2' as mentioned in the new paper?

**Comment by JulienSiems** [comment]
Hi @cptay, 

You are right. I am a little confused why they cite it as if log σ² was used in the Multi Task Paper as regularization. Maybe it is a mistake or potentially the version on arxiv is not up to date, because it hasn't been published yet?

**Comment by JulienSiems** [comment]
@cptay Did you get good results with the modified regularization term from the auxillary loss paper?

**Comment by cptay** [comment]
@JulienSiems Yes, the modified regularization gave me better results, but I can only say it works for my case.

Now I wonder how we can improve performance by enhancing the regularization term in a principled and convincing way?

**Comment by LMdeLiangMi** [comment]
Hi, @JulienSiems 
so, should I initialize a variable s = log σ² as a learnable parameter of the model?

**Comment by JulienSiems** [comment]
Hi @LMdeLiangMi,

I initialized my s = log σ² as 2.0 for all my tasks. And then used the transformation as discussed above. This allows the network to jointly optimize the task loss and the task uncertainty.
Depending on the framework, you may have to add it to the collection of trainable variables.

That's how I understood it from the paper. Would be great if @alexgkendall could verify this for us.

**Comment by NicolaiHarich** [comment]
I think that's completely right @JulienSiems!
In the meantime you can follow this example: https://github.com/yaringal/multi-task-learning-example

**Comment by JulienSiems** [comment]
Great thanks for the link @NicolaiHarich!

**Comment by wandering007** [comment]
@LMdeLiangMi @alexgkendall  I have the same concern. The Normal distribution assumption holds perfectly for L2 loss, what if we want to use L1 loss? The paper also uses L1 loss for depth regression, how do the authors handle the variance estimation?

**Comment by JulienSiems** [comment]
@wandering007 based on the seminar slide posted by @LMdeLiangMi it looks like the authors treated the uncertainty for the L1 equally to the L2 loss, because the depth estimation and instance segmentation task is using an L1 loss.

**Comment by wandering007** [comment]
@JulienSiems Thanks! I also find that the used prior is Laplacian distribution instead of Normal distribution for L1 loss, noted in [What Uncertainties Do We Need in Bayesian Deep Learning for Computer Vision?](https://arxiv.org/pdf/1703.04977.pdf).

**Comment by LMdeLiangMi** [comment]
😂 hi everyone, do you know what is the difference  between epistemic uncertainty and homoscedastic uncertainty?

**Comment by wandering007** [comment]
@LMdeLiangMi It seems that Epistemic Uncertainty and Heteroscedastic Aleatoric Uncertainty is out of the scope of the paper. For Homoscedastic uncertainty you can check (@NicolaiHarich 's recommendation)[https://github.com/alexgkendall/multitaskvision/issues/5#issuecomment-417839722].

**Comment by tariyaz** [comment]
Hi, so i tried using this how it's proposed in the paper (using s=log σ²) and L = 0.5*exp(-s)*L_1 + s ... and it happens exactly what is mentioned in the paper with the modified regularization term, i.e. my total loss gets negative because s gets negative (which isn't that bad per se, at the contrary i expect the loss factor to be >1). But it can get unbalanced so that the total loss gets negative.

With the modified regularization term i have to train for s=σ. In my case s gets 0 so the reg. term gets 0. But with s=0 i also get 0.5*exp(-(log 0²)) which is inf.. I guess that's exactly the reason why we need to train for s=log σ², but then i cannot modify the reg. term and i don't think that taking the absolute term helps, because like i said it should be allowed for the term to get negative.

**Comment by LMdeLiangMi** [comment]
@tariyaz same question ... I also get negative loss values

**Comment by aneesh3108** [comment]
Hi @JulienSiems ,
If you dont mind me asking - which framework are you coding in? 

I"m using Pytorch and having a hard time understand how I can initialize s = 2.0 , while s is supposed to be an output of the network (or maybe I'm looking at it the wrong way?)

**Comment by woaichipinngguo** [comment]
@JulienSiems @cptay @wandering007 @alexgkendall 
Is this code snippet right? I think there is something wrong . I got negative loss.
```
myNet=myNet.to(device)
sig1=torch.tensor(2.0,requires_grad=True, device=device)
sig2=torch.tensor(2.0,requires_grad=True, device=device)
criterion=nn.CrossEntropyLoss()

optimizer=optim.Adam([{'params':myNet.parameters()},
                         {'params': [s1,s2]}],lr=lr,weight_decay = opt.weight_decay))
scheduler = optim.lr_scheduler.ReduceLROnPlateau(optimizer, 'min',patience=10)


optimizer.zero_grad()

outline,outvp=myNet(data)
lossline=torch.exp(-s1)*criterion(outline,mask)+(s1/2)
lossvp=torch.exp(-s2)*criterion(outvp,vpmask)+(s2/2)
 loss=lossline+lossvp

loss.backward()
optimizer.step()


```

**Comment by ignaciorlando** [comment]
Well, I checked the math and it is reasonable to get negative weights using this loss.
If we follow Eq. (10) from the paper and we modify it to use <a href="https://www.codecogs.com/eqnedit.php?latex=s=log(\sigma^2)" target="_blank"><img src="https://latex.codecogs.com/gif.latex?s=log(\sigma^2)" title="s=log(\sigma^2)" /></a>, we get:

<a href="https://www.codecogs.com/eqnedit.php?latex=\mathcal{L}&space;=&space;\frac{1}{2}&space;\exp(-s_1)&space;\mathcal{L}_1&space;&plus;&space;\exp(-s_2)&space;\mathcal{L}_2&space;&plus;&space;\frac{s_1}{2}&space;&plus;&space;\frac{s_2}{2}" target="_blank"><img src="https://latex.codecogs.com/gif.latex?\mathcal{L}&space;=&space;\frac{1}{2}&space;\exp(-s_1)&space;\mathcal{L}_1&space;&plus;&space;\exp(-s_2)&space;\mathcal{L}_2&space;&plus;&space;\frac{s_1}{2}&space;&plus;&space;\frac{s_2}{2}" title="\mathcal{L} = \frac{1}{2} \exp(-s_1) \mathcal{L}_1 + \exp(-s_2) \mathcal{L}_2 + \frac{s_1}{2} + \frac{s_2}{2}" /></a>

<a href="https://www.codecogs.com/eqnedit.php?latex=\exp(-s_1)" target="_blank"><img src="https://latex.codecogs.com/gif.latex?\exp(-s_1)" title="\exp(-s_1)" /></a> is always positive for every <a href="https://www.codecogs.com/eqnedit.php?latex=s" target="_blank"><img src="https://latex.codecogs.com/gif.latex?s" title="s" /></a> value, so in principle this ensures to have positive <a href="https://www.codecogs.com/eqnedit.php?latex=\sigma" target="_blank"><img src="https://latex.codecogs.com/gif.latex?\sigma" title="\sigma" /></a> estimates. However, when <a href="https://www.codecogs.com/eqnedit.php?latex=s&space;>&space;0" target="_blank"><img src="https://latex.codecogs.com/gif.latex?s&space;>&space;0" title="s > 0" /></a>, <a href="https://www.codecogs.com/eqnedit.php?latex=\exp(-s_1)" target="_blank"><img src="https://latex.codecogs.com/gif.latex?\exp(-s_1)" title="\exp(-s_1)" /></a> gets (really) high values, and of course the regularizer gets high as well. On the other hand, when <a href="https://www.codecogs.com/eqnedit.php?latex=s&space;>&space;0" target="_blank"><img src="https://latex.codecogs.com/gif.latex?s&space;>&space;0" title="s << 0" /></a>, then <a href="https://www.codecogs.com/eqnedit.php?latex=\exp(-s_1)" target="_blank"><img src="https://latex.codecogs.com/gif.latex?\exp(-s_1)" title="\exp(-s_1)" /></a> tends to 0, and the negative regularizer may cause the loss to be negative. So yes, with this regularizer we cannot prevent the weights to be negative.

**Comment by woaichipinngguo** [comment]
@ignaciorlando 
Yes. Negative loss is normal situation. The negative training  loss  also occurs in the paper.
![111](https://user-images.githubusercontent.com/15664013/47723057-b0887600-dc8e-11e8-9757-d0c35b229ae9.jpg)

The weight for loss is:
![112](https://user-images.githubusercontent.com/15664013/47723377-5c31c600-dc8f-11e8-8ff0-cae60a594b91.JPG)
so weifht is larger than zero.

**Comment by ignaciorlando** [comment]
Hi @woaichipinngguo. 

From what I got from Fig. 6 on the paper, the values that are represented in those plots are the <a href="https://www.codecogs.com/eqnedit.php?latex=\exp(-s)" target="_blank"><img src="https://latex.codecogs.com/gif.latex?s" title="s" /></a> values for each task, right? Not the loss values. But having negatives <a href="https://www.codecogs.com/eqnedit.php?latex=\exp(-s)" target="_blank"><img src="https://latex.codecogs.com/gif.latex?s" title="s" /></a> values can cause the loss to become negative. However, at the top row of Fig. 7 we can see that each individual loss remains positive. It would be interesting to see if summing up these losses with the actual s values still results in a positive loss, right? 

Of course each weight is still positive, sorry for my wording in my last post. I wanted to mention that the s values could be negative, not the weights. Thanks for pointing that out!

**Comment by SergeLiu** [comment]
@ignaciorlando I've used this uncertainty to optimize the loss metioned in Alex's paper [_Geometric loss functions for camera pose regression with deep learning_](http://cn.arxiv.org/pdf/1704.00390v2),and I got the negative loss too.
![image](https://user-images.githubusercontent.com/40231580/48300974-345d2080-e521-11e8-8828-6dd03b8d4be2.png)
as shown function in the figure,S_x and S_q can be negative so loss can't converge to lead a satisfied result. Have you tried the new reg term in new paper as  https://github.com/alexgkendall/multitaskvision/issues/5#issuecomment-412485851 mentioned?

**Comment by woaichipinngguo** [comment]
> hi, @JulienSiems Thanks for your reply! I find that in the paper, the author uses MSE, right?
> ![image](https://user-images.githubusercontent.com/13929748/44960379-99b2e680-aecc-11e8-880a-9943c167bbaa.png)
> But how should I modify this equation if I want to use RMSE? just y1-fw(x)

what the difference between y and fW(x)?   In he paper ,  fW(x) is network output and y is model output. I don't understand. where is ground-truth?
@ignaciorlando  @wandering007  @JulienSiems

**Comment by woaichipinngguo** [comment]
![1](https://user-images.githubusercontent.com/15664013/48845482-934f4f00-edd7-11e8-82fc-ab0aaddc96a3.JPG)
![2](https://user-images.githubusercontent.com/15664013/48845487-96e2d600-edd7-11e8-8625-f975b567b35b.JPG)
@alexgkendall @ignaciorlando @JulienSiems @cptay

**Comment by Mikoto10032** [comment]
I have the same problem that the network finally gets negative loss. I have tried two different formulate to weigh multi-task loss but they failed. So I hope someone can deal this problem, thanks in advance!

**Comment by SergeLiu** [comment]
@Mikoto10032 把σ初始化了，改成log(1+σ^2）

**Comment by SergeLiu** [comment]
@Mikoto10032 把σ初始化了，改成log(1+σ^2),御坂10031如是说道

**Comment by Mikoto10032** [comment]
> @Mikoto10032 把σ初始化了，改成log(1+σ^2),御坂10031如是说道

是, 多谢Mikoto10031, 我这就尝试一下

**Comment by LMdeLiangMi** [comment]
Referring to this https://github.com/yaringal/multi-task-learning-example/blob/master/multi-task-learning-example.ipynb, I found that if there are two losses l1 and l2 for taks1 and task2, and l1 is smaller, then the noise σ of task1 will be smaller and then the weight of task1 1/σ^2 is higher. So, I do not know whether I am wrong, we pay more attention to the smaller loss? Like using loss1*loss2 as the final loss? https://piccolboni.info/2018/03/a-simple-loss-function-for-multi-task-learning-with-keras-implementation.html
